### DISCUSSION TITLE
How can I make data from server req available to getInitialProps in both client and server

### DISCUSSION THREAD

**Discussion by carystanley** [vercel/next.js#47125] [Help]
### Summary

For each page I make data requests in `getInitialProps` both on the client and the server using an isomorphic data transport (fetch on client and a request library with additional logging etc.. on the server side)

I server the Next.js through express, I have a number of middleware setup in express to determine if the user is logged in, what country they are from, what the environment is dev, prod, or staging, etc.., language preferences, timezone, etc..

I can take the data from req in the a Custom App and memoize in clientside so that it can be available on the context (feels hacky but works):

```
import { useMemo } from 'react';
import App from 'next/app'
import SessionContext from '../components/context/SessionContext';

function MyApp({ Component, session: initialSession, pageProps }) {
    const session = useMemo(() => (initialSession), []);

    return (
        <SessionContext.Provider value={session}>
            <Component {...pageProps} />
        </SessionContext.Provider>
    );
}

MyApp.getInitialProps = async (appContext) => {
    const { ctx } = appContext;
    const appProps = await App.getInitialProps(appContext);

    return {
        ...appProps,
        session: {
            context: ctx?.req?.context,
            user: ctx?.req?.user
        }
    };
}

export default MyApp
```

 But how can I store these values client-side and make them available in the Page `getInitialProps`??

```
ExamplePage.getInitialProps = (context) => {
    const { country } = context;
    return promiseAllMap({
        redzone: loadData(context, { country })
    });
};
```

I want `loadData` to be able to adjust the host of API calls based on environment (dev, stage, prod), and also be able to make different requests based on the "context" I pulled from the `req` (and have it work both on client & server)

Is there a way to pull information from the req object and make it available on both client and server in the the "context" padded to Page `getInitialProps`?

**Comment by icyJoseph** [comment]
The official go to solution, now seems to be the `appDir`. 

Have you looked at the contents of `__NEXT_DATA__`, it is a script tag with `application/json` content type, which you could read in the client? Though this is not bulletproof I think, but you could fallback to that whenever there's SPA transitions that run completely client side.

In addition to potentially falling stale, the above would also block the main thread when parsing the json-string.

It is a ugly hack, but it might do the trick for you.

**Comment by carystanley** [comment]
@icyJoseph thanks for the reply! You can look into the contexts of the `__NEXT_DATA__` script tag.

In fact Next.js already parses it, and you can access it client-side as `window.__NEXT_DATA__`

If I do that I can get it to work with the following logic changes to `_app.js`

```
const getSession = (ctx={}) => (
    ctx.req ?
        {
            context: ctx?.req?.context,
            user: ctx?.req?.user
        }
        : (typeof window !== 'undefined') && window?.__NEXT_DATA__?.props.session
);


MyApp.getInitialProps = async (appContext) => {
    const { ctx } = appContext;
    const session = getSession(ctx);

    const extendedAppContext = {
        ...appContext,
        ctx: {
            ...ctx,
            session
        }
    }
    const appProps = await App.getInitialProps(extendedAppContext);

    return {
        ...appProps,
        session
    };
}
```

But scared that I am accessing "internal" parts of Next.js that could later change.
So hoping there is a better answer out there that avoids playing with Next.js "internals"

**Reply by icyJoseph** [reply]
Yeah, and accessing it from the window is a non starter for me. I rather pay the serialization price.

Though that one ain't going away anytime soon. I think, the main focus will be the appDir.
